app/package.json

```diff
@@ -35,6 +35,7 @@
     "fs-extra": "^6.0.0",
     "fuzzaldrin-plus": "^0.6.0",
     "keytar": "^4.4.1",
+    "mem": "^4.3.0",
```

Comment:
I think it's hard to give good general advice on patterns where memoization adds significant value.  The general advice around caching and optimizations are always sound, measure before, understand the impact, and implement optimizations where there's demonstrable user-noticeable gains in doing so. It's so easy to get fixated on improving a particular number instead of looking at the user impact. In that vein I think it's wise to (where possible) profile using real world data. In this PR I looked at a repository with no open PR, desktop/desktop (which should be considered upper medium to high in terms of PR velocity) and an internal extremely high activity repository and looked at the cost for each with the end goal being showing PRs to the user faster and with less latency from the server.

I did write a little bit more about a specialized version of memoization in https://github.com/desktop/desktop/pull/6278 which is the [pattern recommended by React](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#what-about-memoization). That pattern has worked well I feel and is relatively safe to use since it cannot grow unbounded memory-wise.

> Or that we should _avoid_ memoization?

I think this is a little easier to answer, at least when we're referring to this particular library. When I see `mem(function)` what I look for is whether the internal Map for that is unbounded in growth. In other words, memoizing things like email addresses, commit shas, etc would make me concerned and prompt me to dig deeper. If we end up running into such cases I'd recommend pairing this library with something like `quick-lru` to safe-guard against unbounded memory growth.

In the two cases in this PR where I've used this memoization library the internal map gets garbage collected at the end of the closure so there's no risk for unbounded growth there and the in the trade-off between memory and cpu there's no debate given that the CPU route is so much more time consuming.

I'd also recommend understanding how each of these memoization functions deals with comparing argument values for equality. The `mem` library, for example, when called with more than one argument will use JSON.stringify on the arguments array use that string as the key. This may or may not be what you want. In this PR there's actually a good case where we don't want that:

```ts
const upsertRepo = mem(store.upsertGitHubRepository.bind(store), {
  cacheKey: (_, repo) => repo.clone_url,
})
```

In this case the `repo` argument is of type `IAPIRepository` which seems innocuous enough but that we know is simply a coerced type and is actually a complex argument that could (and do) consist of much more information than we have modelled in the interface. Therefore I've specifically told the memoization library what to use as the cache key here.

Reply:
@niik this is very helpful - thanks!